apis/ingress/v1alpha1/astartedefaultingress_webhook.go

```diff
@@ -0,0 +1,248 @@
+		r.Name,
+		allErrors,
+	)
+}
+
+func (r *AstarteDefaultIngress) validateBrokerServiceType() *field.Error {
+	if pointy.BoolValue(r.Spec.Broker.Deploy, true) {
+		if r.Spec.Broker.ServiceType == v1.ServiceTypeNodePort || r.Spec.Broker.ServiceType == v1.ServiceTypeLoadBalancer {
+			return nil
+		}
+
+		fldPath := field.NewPath("spec").Child("broker").Child("serviceType")
+		err := errors.New("Wrong broker service type. Allowed values: LoadBalancer, NodePort.")
+
+		astartedefaultingresslog.Error(err, "Allowed service types for the Broker are: LoadBalancer and NodePort")
+		return field.Invalid(fldPath, r.Spec.Broker.ServiceType, err.Error())
+	}
+	return nil
+}
+
+func (r *AstarteDefaultIngress) validateReferencedAstarte(c client.Client) (*apiv1alpha1.Astarte, *field.Error) {
+	fldPath := field.NewPath("spec").Child("astarte")
+
+	// ensure that the referenced Astarte instance exists
+	theAstarte := &apiv1alpha1.Astarte{}
+	if err := c.Get(context.Background(), types.NamespacedName{Name: r.Spec.Astarte, Namespace: r.Namespace}, theAstarte); err != nil {
+		astartedefaultingresslog.Error(err, "The referenced Astarte does not exist.")
+		return nil, field.Invalid(fldPath, r.Spec.Astarte, err.Error())
+	}
+	return theAstarte, nil
+}
+
+func (r *AstarteDefaultIngress) validateBrokerTLSConfig(astarte *apiv1alpha1.Astarte) *field.Error {
+	if pointy.BoolValue(astarte.Spec.VerneMQ.SSLListener, false) && astarte.Spec.VerneMQ.SSLListenerCertSecretName == "" {
+		err := errors.New("Broker TLS is misconfigured. Review your Astarte CR to ensure TLS termination at VerneMQ level.")
+		fldPath := field.NewPath("astarte").Child("spec").Child("vernemq").Child("sslListenerCertSecretName")
+
+		return field.Invalid(fldPath, astarte.Spec.VerneMQ.SSLListenerCertSecretName, err.Error())
+	}
+
+	return nil
+}
+
+func (r *AstarteDefaultIngress) validateDashboardTLSConfig() *field.Error {
+	if r.Spec.TLSSecret == "" && pointy.BoolValue(r.Spec.Dashboard.SSL, true) &&
+		pointy.BoolValue(r.Spec.Dashboard.Deploy, true) && r.Spec.Dashboard.TLSSecret == "" {
+
+		fldPath := field.NewPath("spec").Child("dashboard").Child("tlsSecret")
+		err := errors.New("TLS misconfigured for dashboard.")
+
+		astartedefaultingresslog.Error(err, "Ensure to provide a TLS secret to secure your connection.")
+		return field.Invalid(fldPath, r.Spec.Dashboard.TLSSecret, err.Error())
+	}
+	return nil
+}
+
+func (r *AstarteDefaultIngress) validateAPITLSConfig() *field.Error {
+	if r.Spec.TLSSecret == "" && r.Spec.API.TLSSecret == "" {
+		fldPath := field.NewPath("spec").Child("api").Child("tlsSecret")
+		err := errors.New("TLS misconfigured for API.")
+
+		astartedefaultingresslog.Error(err, "Ensure to provide a TLS secret to secure your connection.")
+		return field.Invalid(fldPath, r.Spec.API.TLSSecret, err.Error())
+	}
+	return nil
+}
+
+func (r *AstarteDefaultIngress) validateRootTLSSecretExistence(c client.Client) *field.Error {
```

Comment:
I think all of these codepaths are extremely verbose and confusing. You want to validate all of the secrets that have been explicitly specified, and I'm sure there's a more concise and less confusing way to do this than having 3 separate functions with useless validation checks